docs/proposals/network-policy.md

```diff
@@ -0,0 +1,259 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# NetworkPolicy
+
+## Abstract
+
+A proposal for implementing a new resource - NetworkPolicy - which
+will enable definition of ingress policies for selections of pods.
+
+The design for this proposal has been created by, and discussed
+extensively within the Kubernetes networking SIG.  It has been implemented
+and tested using Kubernetes API extensions by various networking solutions already.
+
+In this design, users can create various NetworkPolicy objects which select groups of pods and
+define how those pods should be allowed to communicate with each other.  The
+implementation of that policy at the network layer is left up to the
+chosen networking solution.
+
+> Note that this proposal does not yet include egress / cidr-based policy, which is still actively undergoing discussion in the SIG.  In general,
+egress is expected to augment this proposal rather than modify it, with one exception (noted below).
+
+## Implementation
+
+The implmentation in Kubernetes consists of:
+- A v1beta NetworkPolicy API object
+- A field on the `Namespace` object to control isolation, to be developed as an annotation for now, called `net.beta.kubernetes.io/network-isolation`.
+
+### Namespace changes
+
+> Note that this section is subject to some change with the addition of egress policy as mentioned above.
+
+The `Namespace` object can be augmented with the annotation `net.beta.kubernetes.io/network-isolation` which will take one of the following values:
```

Comment:
I have expounded at greater length about this in email, see https://groups.google.com/forum/#!topic/kubernetes-sig-network/K60pIcr8mxI

The reason we need explicit opt-in is exactly because that means to set the default to the opposite of the old (non-extended) behavior.

We can plan that every time we add new subject matter for network policy that requires explicit opt-in we will double the number of allowed values for the one annotation.  But I do not think that scales well.  However, we do not have to be purists about this.  We can start with one annotation that allows four values.  But at the next step of evolution I do not want to go to eight values, and I really do not want to go to 16 after that.